### DISCUSSION TITLE
Using priv field in struct rpmsg_endpoint for storing application data

### DISCUSSION THREAD

**Discussion by mkaricheri** [OpenAMP/open-amp#546] [General]
```
/**
 * struct rpmsg_endpoint - binds a local rpmsg address to its user
 * @name: name of the service supported
 * @rdev: pointer to the rpmsg device
 * @addr: local address of the endpoint
 * @dest_addr: address of the default remote endpoint binded.
 * @cb: user rx callback, return value of this callback is reserved
 *      for future use, for now, only allow RPMSG_SUCCESS as return value.
 * @ns_unbind_cb: end point service unbind callback, called when remote
 *                ept is destroyed.
 * @node: end point node.
 * @priv: private data for the driver's use
 *
 * In essence, an rpmsg endpoint represents a listener on the rpmsg bus, as
 * it binds an rpmsg address with an rx callback handler.
 */
struct rpmsg_endpoint {
	char name[RPMSG_NAME_SIZE];
	struct rpmsg_device *rdev;
	uint32_t addr;
	uint32_t dest_addr;
	rpmsg_ept_cb cb;
	rpmsg_ns_unbind_cb ns_unbind_cb;
	struct metal_list node;
	void *priv;
};
```
We are using Open-AMP on Bare-metal (master) with Zephyr (slave). We use multi-endpoints, one per application to allow IPC between applications on the two cores. We have a CPP IPC interface / API to create endpoints. send/receive message. One thing we would like to have is to pass the index of the application endpoint to the RPMSG so that it gets passed to the received message handle to the endpoint. So bumped on to the priv structure field in struct rpmsg_endpoint. It works for us, but I am paranoid about the documentation in the header '@priv: private data for the driver's use'. I see that this is not used by RPMSG to store anything and is passed to the application as part of the message callback

rpmsg_virtio.c
			status = ept->cb(ept, RPMSG_LOCATE_DATA(rp_hdr),
					 rp_hdr->len, rp_hdr->src, ept->priv);

1) Is this intended to be used for application use or driver use?
2) If latter, why is it passed to application?

I think this is a documentation mistake. Could someone clarify this please?
If not used, can we use it for application use?

We maintain a table of endpoints and save the index of the endpoint in priv so that when the core receives a response, we can use the index to identify the endpoint for this response. Otherwise we need to maintain some kind of map or loop through the endpoint names to find the correct endpoint for the response. We also use interrupt, so the loop and string comparison inside the ISR is not preferable.

**Comment by mkaricheri** [comment]
I see a discussion here related to this and I am evaluating it. https://github.com/OpenAMP/open-amp/discussions/296 but would appreciate what is the intended use of the priv field.

**Comment by arnopo** [comment]
Hello @mkaricheri 
You can have a look to the [openamp_rsc_table] (https://elixir.bootlin.com/zephyr/latest/source/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c) sample.
It provides an example of the use of the priv field for the endpoint.
Here is the code related to the init part of the endpoint:
https://elixir.bootlin.com/zephyr/latest/source/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c#L312

```c
	tty_ept.priv = &tty_msg;
	ret = rpmsg_create_ept(&tty_ept, rpdev, "rpmsg-tty",
			       RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
			       rpmsg_recv_tty_callback, NULL);
```
